src/solana/SolanaEngine.ts

```diff
@@ -838,16 +851,82 @@ export class SolanaEngine extends CurrencyEngine<
   ): Promise<EdgeTransaction> {
     if (edgeTransaction.signedTx == null) throw new Error('Missing signedTx')
 
+    const { blockhash, lastValidBlockHeight } = asSolanaTxOtherParams(
+      getOtherParams(edgeTransaction)
+    )
     const stakedConnections = this.tools.makeConnections(
       this.networkInfo.stakedConnectionRpcNodes
     )
-    const connections = [...this.tools.connections, ...stakedConnections]
+    const rpcConnections = this.tools.connections
+    const allConnections = [...this.tools.connections, ...stakedConnections]
+
+    const confirmationController = new AbortController()
+    const retryTxController = new AbortController()
+
+    const submitTx = async (): Promise<string> => {
+      const broadcastPromises = allConnections.map(async connection => {
+        const txid = await connection.sendEncodedTransaction(
+          edgeTransaction.signedTx,
+          { skipPreflight: true }
+        )
+        return txid
+      })
+      const txid = await promiseAny(broadcastPromises)
+      return txid
+    }
+
+    const checkBlockheight = async (): Promise<number> => {
+      const broadcastPromises = rpcConnections.map(async connection => {
+        const blockheight = await timeout(
+          connection.getBlockHeight('confirmed'),
+          1000
+        )
+        return blockheight
+      })
+      const blockheightResults = await Promise.allSettled(broadcastPromises)
+      const blockheights = blockheightResults
+        .filter(p => p.status === 'fulfilled')
+        .map(r => (r as PromiseFulfilledResult<number>).value)
+      const maxHeight = Math.max(...blockheights)
+      return maxHeight
+    }
+
+    const retryTxSubmission = async (): Promise<void> => {
+      while (true && !retryTxController.signal.aborted) {
+        const height = await checkBlockheight()
+        if (height < lastValidBlockHeight) {
+          await submitTx()
+        } else {
+          confirmationController.abort()
+          throw new Error('transaction expired')
+        }
+      }
+    }
 
     try {
-      const promises = connections.map(async connection => {
-        return await connection.sendEncodedTransaction(edgeTransaction.signedTx)
+      const txid = await submitTx()
+
+      const confirmPromises = rpcConnections.map(async connection => {
+        const confirmTxRes = await connection.confirmTransaction(
+          {
+            signature: txid,
+            blockhash,
+            lastValidBlockHeight,
+            abortSignal: confirmationController.signal
+          },
+          'confirmed'
+        )
+        const txError = asMaybe(asString)(confirmTxRes.value.err)
+        if (txError != null) {
+          throw new Error(txError)
+        }
+
+        // Confirmed!!
+        retryTxController.abort()
       })
-      const txid: TransactionSignature = await promiseAny(promises)
+
+      await Promise.race([retryTxSubmission(), ...confirmPromises])
+
       edgeTransaction.txid = txid
```

Comment:
Where does `txid` come from? Both `confirmPromises` and `submitTx` appear to use `const txid` locally, so I'm not sure how it flows out into this outer scope.

Reply:
It's returned from the standalone call to submitTx up above